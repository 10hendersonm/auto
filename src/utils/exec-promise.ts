import * as childProcess from 'child_process';
import { promisify } from 'util';

const exec = promisify(childProcess.exec);

export default async function execPromise(cmd: string) {
  const { stdout, stderr } = await exec(cmd, { maxBuffer: 1024 * 1000 });

  if (stderr) {
    console.log(stderr);
  }

  return stdout.trim();
}
