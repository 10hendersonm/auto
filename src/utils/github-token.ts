import fs from 'fs';
import path from 'path';
import registryUrl from 'registry-url';
import { promisify } from 'util';
import settingsUrl from './settings-url';

const readFile = promisify(fs.readFile);
const registry = registryUrl();
// tslint:disable-next-line
const normalizedRegistry = registry.replace('http:', '').replace('https:', '');

export default async function getGitHubToken(
  apiUrl = 'https://api.github.com'
): Promise<string> {
  if (process.env.GH_TOKEN) {
    return process.env.GH_TOKEN;
  }

  const helpText = `Try setting an access token up ${settingsUrl(apiUrl)}`;

  if (!process.env.HOME) {
    throw new Error(`Can't find the GH_TOKEN and no HOME defined. ${helpText}`);
  }

  const rcLocation = path.resolve(process.env.HOME, '.npmrc');

  if (fs.existsSync(rcLocation)) {
    const rc = await readFile(rcLocation, 'utf8');

    const regex = new RegExp(`${normalizedRegistry}:_authToken=\(\\w+\)`);
    const token = rc.match(regex);
    if (token && token[1]) {
      return token[1];
    }

    throw new Error(`No token in the .npmrc. ${helpText}`);
  }

  throw new Error(`Can't find a GitHub token to use. ${helpText}`);
}
