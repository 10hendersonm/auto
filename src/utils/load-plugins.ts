import * as path from 'path';
import { AutoRelease } from '../main';
import NPMPlugin from '../plugins/npm';

export type IPluginConstructor = new () => IPlugin;

export interface IPlugin {
  name: string;
  apply(auto: AutoRelease): void;
}

type SupportedPlugin = 'npm';
const plugins = new Map<SupportedPlugin, IPluginConstructor>([
  ['npm', NPMPlugin]
]);

function isSupported(key: SupportedPlugin | string): key is SupportedPlugin {
  return !!plugins.get(key as SupportedPlugin);
}

export default function loadPlugin(
  pluginPath: SupportedPlugin | string
): IPluginConstructor | undefined {
  if (isSupported(pluginPath)) {
    return plugins.get(pluginPath);
  }

  try {
    return require(pluginPath);
  } catch (error) {
    return require(path.join(process.cwd(), pluginPath));
  }
}
