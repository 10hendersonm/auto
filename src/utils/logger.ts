import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
