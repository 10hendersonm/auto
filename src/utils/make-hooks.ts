import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
