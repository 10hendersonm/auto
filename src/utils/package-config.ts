import fs from 'fs';
import parseGithubUrl from 'parse-github-url';
import { promisify } from 'util';

export interface IRepoConfig {
  owner: string;
  repo: string;
  version: string;
}

interface IPackageConfig {
  version: string;
  repository?:
    | {
        url: string;
      }
    | 'string';
}

const readFile = promisify(fs.readFile);

const getPackageConfig = async (): Promise<IPackageConfig> => {
  const pkgConfig = await readFile('./package.json', 'utf-8');
  return JSON.parse(pkgConfig);
};

export default async function getConfigFromPackageJson(): Promise<IRepoConfig> {
  const { repository, version } = await getPackageConfig();

  if (!repository) {
    throw new Error('Cannot read repo info from package.json');
  }

  const { owner, name } =
    parseGithubUrl(
      typeof repository === 'string' ? repository : repository.url
    ) || ({} as any);

  if (!owner || !name) {
    throw new Error(
      'Cannot read owner and package name from Github URL in package.json'
    );
  }

  return {
    repo: name,
    owner,
    version
  };
}
