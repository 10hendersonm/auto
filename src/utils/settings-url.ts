import * as url from 'url';

export default function settingsUrl(apiUrl: string) {
  let { host } = url.parse(apiUrl);

  if (host && host.includes('api.')) {
    host = host.replace('api.', '');
  }

  return `https://${host}/settings/tokens/new`;
}
