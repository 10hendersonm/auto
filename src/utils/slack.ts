import fetch from 'node-fetch';
import join from 'url-join';

const sanitizeMarkdown = (markdown: string) =>
  markdown
    .split('\n')
    .map(line => line.replace(/\[([\s\S]+)\]\([\S\s]+\)/, (_, label) => label))
    .join('\n');

interface IPostToSlackOptions {
  tag: string;
  owner: string;
  repo: string;
  slackUrl: string;
  baseUrl: string;
}

export default async function postToSlack(
  releaseNotes: string,
  options: IPostToSlackOptions
) {
  const body = `\`\`\`${sanitizeMarkdown(releaseNotes)}\`\`\``;
  const token = process.env.SLACK_TOKEN;
  const url = `${join(
    options.baseUrl,
    options.owner,
    options.repo,
    'releases/tag',
    options.tag
  )}\n`;

  if (!token) {
    return Promise.reject(new Error('Slack needs a token to send a message'));
  }

  return fetch(`${options.slackUrl}?token=${token}`, {
    method: 'POST',
    body: JSON.stringify({
      text: [`@channel: New release *${options.tag}*`, url, body].join('\n'),
      // eslint-disable-next-line camelcase
      link_names: 1
    }),
    headers: { 'Content-Type': 'application/json' }
  });
}
