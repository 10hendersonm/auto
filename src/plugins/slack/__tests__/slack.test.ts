import SlackPlugin from '..';
import Auto from '../../../auto';
import { dummyLog } from '../../../utils/logger';

const fetchSpy = jest.fn();
// @ts-ignore
jest.mock('node-fetch', () => (...args) => {
  fetchSpy(...args);
});

beforeEach(() => {
  fetchSpy.mockClear();
});

const mockAuto = ({
  git: {
    options: {
      owner: 'Adam Dierkens',
      repo: 'test'
    },
    getProject: () => ({
      html_url: 'https://github.custom.com'
    })
  },
  logger: dummyLog()
} as unknown) as Auto;

describe('postToSlack', () => {
  test('should throw without a token', async () => {
    const plugin = new SlackPlugin({ url: 'https://custom-slack-url' });

    await expect(
      plugin.postToSlack(mockAuto, 'v1.2.3', '# My Notes')
    ).rejects.toEqual(new Error('Slack needs a token to send a message'));
  });

  test('should call slack api', async () => {
    const plugin = new SlackPlugin({ url: 'https://custom-slack-url' });
    process.env.SLACK_TOKEN = 'MY_TOKEN';

    await plugin.postToSlack(
      mockAuto,
      '1.0.0',
      '# My Notes\n- PR [some link](google.com)'
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      'https://custom-slack-url?token=MY_TOKEN'
    );
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });
});
