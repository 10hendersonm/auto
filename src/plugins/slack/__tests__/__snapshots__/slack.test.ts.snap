// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postToSlack should call slack api 1`] = `"{\\"text\\":\\"@channel: New release *<https://github.custom.com/Adam Dierkens/test/releases/tag/1.0.0|1.0.0>*\\\\n* My Notes*\\\\nâ€¢ PR <google.com|some link>\\",\\"link_names\\":1}"`;
